File: js/1976.js
const fs = require('fs');

const inputs = fs.readFileSync('example.txt').toString().trim().split('\n');

const N = +inputs[0];
const M = +inputs[1];
const parent = Array(N)
  .fill()
  .map((_, idx) => idx);

for (let i = 2; i < N + 2; i++) {
  inputs[i].split(' ').forEach((e, j) => {
    if (+e === 1 && !sameParent(i - 1, j + 1)) {
      unionParent(i - 1, j + 1);
    }
  });
}

const plan = inputs[N + 2].split(' ').map(Number);

solve();

function solve() {
  for (let i = 0; i < M - 1; i++) {
    const from = plan[i];
    const to = plan[i + 1];
    if (!sameParent(from, to)) {
      console.log('NO');
      return;
    }
  }
  console.log('YES');
}

function getParent(x) {
  if (x === parent[x]) return x;
  else return (parent[x] = getParent(parent[x]));
}

function unionParent(a, b) {
  a = getParent(a);
  b = getParent(b);
  if (a < b) parent[b] = a;
  else parent[a] = b;
}

function sameParent(a, b) {
  a = getParent(a);
  b = getParent(b);
  return a === b;
}

File: c++/1976.cpp
#include <iostream>
#include <vector>

using namespace std;

int N, M;
vector<int> plan;
int parent[201];

int getParent(int x) {
  if (x == parent[x])
    return x;
  else
    return parent[x] = getParent(parent[x]);
}

void unionParent(int a, int b) {
  a = getParent(a);
  b = getParent(b);
  if (a > b)
    parent[a] = b;
  else
    parent[b] = a;
}

bool sameParent(int a, int b) {
  a = getParent(a);
  b = getParent(b);
  return a == b;
}

int main() {
  freopen("example.txt", "r", stdin);
  cin >> N >> M;

  for (int i = 1; i <= N; i++) {
    parent[i] = i;
  }

  for (int i = 1; i <= N; i++) {
    for (int j = 1; j <= N; j++) {
      int connected;
      cin >> connected;
      if (connected == 1 && !sameParent(i, j)) unionParent(i, j);
    }
  }

  for (int i = 0; i < M; i++) {
    int a;
    cin >> a;
    plan.push_back(a);
  }

  for (int i = 0; i < M - 1; i++) {
    int from = plan[i];
    int to = plan[i + 1];
    if (!sameParent(from, to)) {
      cout << "NO";
      return 0;
    }
  }

  cout << "YES";

  return 0;
}
